src/test/java/org/jenkins/plugins/lockableresources/LockStepTest.java

```diff
@@ -513,4 +515,89 @@ public void evaluate() throws Throwable {
 			}
 		});
 	}
+
+	@Issue("JENKINS-40879")
+	@Test
+	public void parallelLockRelease() throws Exception {
+		story.addStep(new Statement() {
+			@Override
+			public void evaluate() throws Throwable {
+				LockableResourcesManager.get().createResource("resource1");
+				LockableResourcesManager.get().createResource("resource2");
+				LockableResourcesManager.get().createResource("resource3");
+				LockableResourcesManager.get().createResource("resource4");
+				WorkflowJob j = story.j.jenkins.createProject(WorkflowJob.class, "j");
+				j.setDefinition(new CpsFlowDefinition(
+						"lock(resource: 'resource1') {\n" +
+								"    sleep 1\n" +
```

Comment:
I am deliberately not using `semaphore` for the early `lock`s (not all of which may be needed - I had some problems getting it reproducible due to the bug only popping up when there's another `context` waiting for the resource getting unlocked) because I specifically want the builds to bludgeon their way through the first three `lock`s, not have the test let them get through one by one. So far as I can tell, it's basically impossible to reproduce the bug if you're explicitly proceeding each `semaphore` one by one.